File: Programmers/Level_1/크레인인형뽑기.py
'''
https://programmers.co.kr/learn/courses/30/lessons/64061

기본적인 스택 문제 + 2차원 배열 탐색 
'''

def solution(board, moves):
    answer = 0
    N = len(board)
    st = []
    
    for move in moves:
        for i in range(N):
            if board[i][move-1] != 0:
                if st and st[-1] == board[i][move-1]:
                    answer += 2
                    st.pop()
                else:
                    st.append(board[i][move-1])
                    
                board[i][move-1] = 0
                break
    
    return answer

print(solution([[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]], [1,5,3,5,1,2,1,4]))

File: Programmers/Level_1/Swift_Codes/Swift_Codes/크레인인형뽑기.swift
//
//  크레인인형뽑기.swift
//  Swift_Codes
//
//  Created by InJe Choi on 2022/05/04.
//

/*
 https://programmers.co.kr/learn/courses/30/lessons/64061
 
 기본적인 스택 문제 + 2차원 배열 탐색
 */

import Foundation

func solution(_ board: [[Int]], _ moves: [Int]) -> Int {
    var board = board
    var answer = 0
    var stack = [Int]()
    let N = board.count
    
    for move in moves {
        for i in 0..<N {
            if board[i][move-1] != 0 {
                if !stack.isEmpty && stack.last! == board[i][move-1] {
                    answer += 2
                    stack.removeLast()
                } else {
                    stack.append(board[i][move-1])
                }
                
                board[i][move-1] = 0
                break
            }
        }
    }
    
    return answer
}
